monai/data/wsi_reader.py

```diff
@@ -0,0 +1,428 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import Any, Dict, List, Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+from monai.config import DtypeLike, PathLike
+from monai.data.image_reader import ImageReader, _stack_images
+from monai.data.utils import is_supported_format
+from monai.transforms.utility.array import EnsureChannelFirst
+from monai.utils import ensure_tuple, optional_import, require_pkg
+
+CuImage, _ = optional_import("cucim", name="CuImage")
+
+__all__ = ["BaseWSIReader", "WSIReader", "CuCIMWSIReader"]
+
+
+class BaseWSIReader(ImageReader):
+    """
+    An abstract class that defines APIs to load patches from whole slide image files.
+
+    Typical usage of a concrete implementation of this class is:
+
+    .. code-block:: python
+
+        image_reader = MyWSIReader()
+        wsi = image_reader.read(, **kwargs)
+        img_data, meta_data = image_reader.get_data(wsi)
+
+    - The `read` call converts an image filename into whole slide image object,
+    - The `get_data` call fetches the image data, as well as meta data.
+
+    The following methods needs to be implemented for any concrete implementation of this class:
+
+    - `read` reads a whole slide image object from a given file
+    - `get_size` returns the size of the whole slide image of a given wsi object at a given level.
+    - `get_level_count` returns the number of levels in the whole slide image
+    - `get_patch` extracts and returns a patch image form the whole slide image
+    - `get_metadata` extracts and returns metadata for a whole slide image and a specific patch.
+
+
+    """
+
+    supported_formats: List[str] = []
```

Comment:
I would suggest to use `supported_ suffixes` to be consistent with other readers and the function name:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L387
And I think you are making a good change to put the `supported_suffixes` as a class member, I may also change all the other readers soon.

Thanks.